Assignment/FSMTest.c

```diff
@@ -0,0 +1,129 @@
+      cmdline,          // Command line (path to the executable)
+      NULL,            // Process handle not inheritable
+      NULL,            // Thread handle not inheritable
+      FALSE,           // Set handle inheritance to FALSE
+      0,               // No creation flags
+      NULL,            // Use parent's environment block
+      NULL,            // Use parent's starting directory
+      &si,             // Pointer to STARTUPINFO structure
+      &pi))            // Pointer to PROCESS_INFORMATION structure
+   {
+      // If CreateProcess fails, print an error message
+      int err = GetLastError ();
+      printf ("Failed to start process. Error: %lu\n", GetLastError ());
+      free (cmdline);
+      return 1;
+   }
+
+   // Wait until the process has finished execution
+   WaitForSingleObject (pi.hProcess, INFINITE);
+
+   // Close handles
+   CloseHandle (pi.hProcess);
+   CloseHandle (pi.hThread);
+
+   free (cmdline);
+   return 0;
+
+}
+
+bool CompareFiles (char* file1, const char* file2) {
+   FILE* f1 = fopen (file1, "r"), * f2 = fopen (file2, "r");
+   if (f1 == NULL || f2 == NULL) {
+      printf ("Error opening files\n");
+      return false;
+   }
+   int pos = 0, ch1, ch2;
+   while ((ch1 = fgetc (f1)) != EOF && (ch2 = fgetc (f2)) != EOF) {
+      if (ch1 != ch2) {
+         printf ("Error at bit no. %d, Expected %c, Actual %c\n", pos, ch1, ch2);
+         fclose (f1);
+         fclose (f2);
+         return false;
+      }
+      pos++;
+   }
+   fclose (f1);
+   fclose (f2);
+   return true;
+}
+
+/// <summary>
+/// Test Harness
+/// </summary>
+/// <param name="argc"></param>
+/// <param name="argv">argv[1] is the name of the FSM</param>
+/// <returns></returns>
+int main (int argc, char** argv) {
+#define NTESTS 7
+   if (argc != 2) {
+      printf ("Usage: %s <FSM executable name>\n", argv[0]);
+      return -1;
+   }
+   const char* inputFiles[] = { "test1in.txt", "test2in.txt", "test3in.txt", "test4in.txt",
+                                "test5in.txt", "test6in.txt", "test7in.txt" },
+      * expectedFiles[] = { "test1ref.txt", "test2ref.txt", "test3ref.txt", "test4ref.txt",
+                         "test5ref.txt", "test6ref.txt", "test7ref.txt" };
+   char* outputFiles[] = { "test1out.txt", "test2out.txt", "test3out.txt", "test4out.txt",
+                           "test5out.txt", "test6out.txt", "test7out.txt" };
+
+   for (int i = 0; i < NTESTS; i++) {
+      const char* inputFile = inputFiles[i],
+         * expectedFile = expectedFiles[i];
+      char *outputFile = outputFiles[i];
+      printf ("Running test %d with input file: %s\n", i + 1, inputFile);
+      if (ExecProgram (argv[1], inputFile, outputFile) != 0) {
+         printf ("Error executing test %d\n", i + 1);
+         continue;
+      }
+      // Compare the output with the expected reference file
+      if (CompareFiles (outputFile, expectedFile)) printf ("Test %d passed\n\n", i + 1);
+      else printf ("Test %d failed\n\n", i + 1);

```

Comment:
Compare all files and print once if the test has passed or failed. If the test has failed, print relevant details and break out of the loop. At that point there is no point in comparing the remaining tests.